chapters/buffer_device_address.adoc

```diff
@@ -0,0 +1,181 @@
+
+[source,swift]
+----
+%x = OpLoad %type %ptr Aligned 16
+OpStore %ptr %obj Aligned 16
+----
+
+This will match with the `VkMemoryRequirements::alignment` used to create the memory
+
+=== Nullptr
+
+SPIR-V has a `OpConstantNull`, but that can't be used with `PhysicalStorageBuffer`. The way around this is to either convert the pointer to an integer with `OpConvertPtrToU` or to a `uvec2` with `OpBitcast`.
+
+Vulkan defines the interger value of `0` to be null (as everyone would hopefully expect!).
+
+=== Cross stage variables
+
+There is still on-going discussion to clarify if/how passing pointers from two stages works (ex. vertex to fragment).
+
+The main issue is you may get validation layers errors with the `Location` matching (see https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/5349).
+
+The suggestion is to just pass a `uvec2` or `int64` across stages and re-cast it in your consumer stage.
+
+=== OpTypeForwardPointer and infinite loops
+
+The `OpTypeForwardPointer` is used to forward reference the type of the pointer. This is useful if the app wants to do something like a linked-list
+
+[source,glsl]
+----
+layout(buffer_reference) buffer Node;
+
+layout(buffer_reference, std430) buffer Node {
+    Node next_node;
+    int payload;
+};
+
+layout(set = 0, binding = 0, std430) buffer SSBO {
+    Node start;
+};
+----
+
+You will see the following SPIR-V
+
+[source,swift]
+----
+            OpTypeForwardPointer %Node_ptr PhysicalStorageBuffer
+    %SSBO = OpTypeStruct %Node_ptr
+     %int = OpTypeInt 32 1
+    %Node = OpTypeStruct %Node_ptr %int
+%Node_ptr = OpTypePointer PhysicalStorageBuffer %Node
+%SSBO_ptr = OpTypePointer StorageBuffer %SSBO
+     %var = OpVariable %SSBO_ptr StorageBuffer
+----
+
+When parsing this SPIR-V to do reflection, it is very easy to get into an infinite loop, so be careful.
+
+[NOTE]
+====
+If you want some SPIR-V to test this, look at the `buffer_handle_*.spv` tests in link:https://github.com/KhronosGroup/SPIRV-Reflect/blob/main/tests/glsl[SPIR-V Reflect Tests].
+====
+
+=== Accesses
+
+If you take the following simple GLSL example
+
+[source,glsl]
+----
+#version 450
+#extension GL_EXT_buffer_reference : enable
+
+layout(buffer_reference) buffer BDA {
+    int a;
+};
+
+layout(set=0, binding=0) uniform InData {
+	BDA b;
+};
+
+void main() {
+    b.a = 0;
+}
+----
+
+You will see the following SPIR-V
+
+[source,swift]
+----
+%bda_ptr = OpTypePointer PhysicalStorageBuffer %bda_struct
+%ubo_ptr = OpTypePointer Uniform %bda_ptr
+%int_ptr = OpTypePointer PhysicalStorageBuffer %type_int
+
+ %1 = OpAccessChain %ubo_ptr %3 %int_0
+ %2 = OpLoad %bda_ptr %1
+ %3 = OpAccessChain %int_ptr %2 %int_0
+      OpStore %3 %int_0 Aligned 16
+----
+
+It is easy to think that this `OpLoad` here is dereferencing the pointer before we store into it.
+
+This assumption is wrong, and instead the `OpLoad` is being used to calculate an offset into memory so that the `OpStore` access the right location in memory.
```

Comment:
The OpLoad is loading the pointer from the ubo. The access chain is computing an offset from that pointer.